include/proxy-wasm/context_interface.h

```diff
@@ -0,0 +1,622 @@
+   */
+  virtual std::pair<uint32_t, string_view> getStatus() = 0;
+
+  /**
+   * Get the value of a property.  Some properties are proxy-independent (e.g. ["plugin_root_id"])
+   * while others can be proxy-specific.
+   * @param path is a sequence of strings describing a path to a property.
+   * @param result is a location to write the value of the property.
+   */
+  virtual WasmResult getProperty(string_view path, std::string *result) = 0;
+
+  /**
+   * Set the value of a property.
+   * @param path is a sequence of strings describing a path to a property.
+   * @param value the value to set.  For non-string, non-integral types, the value may be
+   * serialized..
+   */
+  virtual WasmResult setProperty(string_view key, string_view value) = 0;
+};
+
+/**
+ * SharedDataInterface is for shaing data between VMs. In general the VMs may be on different
+ * threads.  Keys can have any format, but good practice would use reverse DNS and namespacing
+ * prefixes to avoid conflicts.
+ */
+struct SharedDataInterface {
+  /**
+   * Get proxy-wide key-value data shared between VMs.
+   * @param key is a proxy-wide key mapping to the shared data value.
+   * @param cas is a number which will be incremented when a data value has been changed.
+   * @param data is a location to store the returned stored 'value' and the corresponding 'cas'
+   * compare-and-swap value which can be used with setSharedData for safe concurrent updates.
+   */
+  virtual WasmResult
+  getSharedData(string_view key, std::pair<std::string /* value */, uint32_t /* cas */> *data) = 0;
+
+  /**
+   * Set a key-value data shared between VMs.
+   * @param key is a proxy-wide key mapping to the shared data value.
+   * @param cas is a compare-and-swap value. If it is zero it is ignored, otherwise it must match
+   * the cas associated with the value.
+   * @param data is a location to store the returned value.
+   */
+  virtual WasmResult setSharedData(string_view key, string_view value, uint32_t cas) = 0;
+}; // namespace proxy_wasm
+
+struct SharedQueueInterface {
+  /**
+   * Register a proxy-wide queue.
+   * @param queue_name is a name for the queue. The queue_name is combined with the vm_id (if any)
+   * to make a unique identifier for the queue.
+   * @param token_ptr a location to store a token corresponding to the queue.
+   */
+  virtual WasmResult registerSharedQueue(string_view queue_name,
+                                         SharedQueueDequeueToken *token_ptr) = 0;
+
+  /**
+   * Get the token for a queue.
+   * @param vm_id is the vm_id of the Plugin of the Root Context which registered the queue.
+   * @param queue_name is a name for the queue. The queue_name is combined with the vm_id (if any)
+   * to make a unique identifier for the queue.
+   * @param token_ptr a location to store a token corresponding to the queue.
+   */
+  virtual WasmResult resolveSharedQueue(string_view vm_id, string_view queue_name,
```

Comment:
low pri: not sure where we landed on this but I still prefer being more explicit here. i.e. `registerSharedWriteQueue` and `lookupSharedReadQueue` or similar.

Reply:
changed to lookup.  Done.